File: a/mergesort.py
from random import randrange
from sys import argv
from time import perf_counter

def merge(arr, left, middle, right):
    L = arr[left:middle+1]
    R = arr[middle+1:right+1]

    i = left
    li = 0
    ri = 0
    while li < len(L) and ri < len(R):
        if L[li] < R[ri]:
            arr[i] = L[li]
            li += 1
            i += 1
            continue
            
        arr[i] = R[ri]
        ri += 1
        i += 1

    while li < len(L):
        arr[i] = L[li]
        li += 1
        i += 1

    while ri < len(R):
        arr[i] = R[ri]
        ri += 1
        i += 1
 
        
def sort(arr, left, right):
    if left < right:
        middle = (right + left) // 2

        sort(arr, left, middle)
        sort(arr, middle + 1, right)

        merge(arr, left, middle, right)
       
    
def mergeSort(arr):
    sort(arr, 0, len(arr) - 1)


def getTestArray(number):
    arr = []
    for i in range(0, number):
        arr.append(randrange(0, 10))

    return arr
    

if __name__ == "__main__":
    number = argv[1]
    arr = getTestArray(int(number))

    #print(arr)

    start = perf_counter()
    mergeSort(arr)
    end = perf_counter()

    #print(arr)
    print(f'Time: {(end - start)}s')

File: a/mergesort.js
const merge = (arr, left, middle, right) => {
  let L = arr.slice(left, middle + 1)
  let R = arr.slice(middle + 1, right + 1)
  
  let i = left
  let j = 0
  let k = 0

  while (j < L.length && k < R.length) {
    if (L[j] < R[k]) {
      arr[i] = L[j]
      j++
      i++
      continue
    }

    arr[i] = R[k]
    k++
    i++
  }

  while (j < L.length) {
    arr[i] = L[j]
    j++
    i++
  }
  
  while (k < R.length) {
    arr[i] = R[k]
    k++
    i++
  }
}

const sort = (arr, left, right) => {
  if (left < right) {
    const middle = Math.floor((left + right) / 2)

    sort(arr, left, middle)
    sort(arr, middle + 1, right)

    merge(arr, left, middle, right)
  }
}

const mergesort = (arr) => {
  sort(arr, 0, arr.length - 1)
}

const getTestArray = (int) => {
  let arr = []
  
  for (let i = 0; i < int; i++) {
    arr.push(Math.floor(Math.random() * Math.floor(10)))  
  }

  return arr
}

const main = () => {
  let arr = getTestArray(parseInt(process.argv[2], 10))

  //console.log(arr)
  console.time('mergesort')

  mergesort(arr)
  
  console.timeEnd('mergesort')
  //console.log(arr)
}

main()
